// @sl-nx/superlogin-next/lib/config/default.config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfig = void 0;
const path_1 = __importDefault(require("path"));
/**
 * These are the default settings that will be used if you don't override them
 * in your config
 */
exports.defaultConfig = {
    security: {
        defaultRoles: ['user'],
        sessionLife: 86400,
        tokenLife: 86400,
        loginOnRegistration: false,
        loginOnPasswordReset: false,
        disabledRoutes: [
            'validate-username',
            'validate-email',
            'unlink',
            'session',
            'forgot-username'
        ]
    },
    local: {
        passwordConstraints: {
            presence: true,
            length: {
                minimum: 8,
                message: 'must be at least 8 characters'
            },
            matches: 'confirmPassword'
        },
        usernameField: 'username',
        passwordField: 'password',
        emailUsername: true,
        emailLogin: true,
        usernameLogin: false,
        uuidLogin: false,
        requireEmailConfirm: true,
        sendConfirmEmail: true,
        requirePasswordOnEmailChange: true,
        sendPasswordChangedEmail: true
    },
    dbServer: {
        protocol: 'http://',
        host: 'localhost:5984',
        designDocDir: path_1.default.join(__dirname, '/designDocs'),
        userDB: 'sl_users',
        couchAuthDB: '_users'
    },
    emails: {
        confirmEmail: {
            subject: 'Please confirm your email',
            template: path_1.default.join(__dirname, '../../templates/email/confirm-email.ejs'),
            format: 'text'
        },
        forgotPassword: {
            subject: 'Your password reset link',
            template: path_1.default.join(__dirname, '../../templates/email/forgot-password.ejs'),
            format: 'text'
        },
        modifiedPassword: {
            subject: 'Your password has been modified',
            template: path_1.default.join(__dirname, '../../templates/email/modified-password.ejs'),
            format: 'text'
        },
        signupExistingEmail: {
            subject: 'You already have registered with us',
            template: path_1.default.join(__dirname, '../../templates/email/signup-email-exists.ejs'),
            format: 'text'
        },
        forgotUsername: {
            subject: 'Your username request',
            template: path_1.default.join(__dirname, '../../templates/email/forgot-username.ejs'),
            format: 'text'
        }
    }
};
