// @sl-nx/superlogin-next/lib/mailer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mailer = void 0;
const ejs_1 = require("ejs");
const fs_1 = require("fs");
const nodemailer_1 = __importDefault(require("nodemailer"));
class Mailer {
    constructor(config) {
        var _a;
        // Initialize the transport mechanism with nodemailer
        this.config = config;
        const customTransport = config.mailer.transport;
        if ((_a = config.testMode) === null || _a === void 0 ? void 0 : _a.noEmail) {
            this.transporter = nodemailer_1.default.createTransport(require('nodemailer-stub-transport')());
        }
        else if (customTransport) {
            this.transporter = nodemailer_1.default.createTransport(customTransport(config.mailer.options));
        }
        else {
            this.transporter = nodemailer_1.default.createTransport(
            // @ts-ignore
            config.mailer.options);
        }
    }
    /**
     * Use the same nodemailer config that Superlogin uses to send out an email.
     * @param templateName the entry under the `emails` property in the config
     * @param recepient the recepient's email address
     * @param data additional data can be passed to `ejs.render()`
     */
    sendEmail(templateName, recepient, data) {
        var _a, _b, _c;
        // load the template and parse it
        let templateFiles = (_a = this.config.emails[templateName]) === null || _a === void 0 ? void 0 : _a.templates;
        if (!templateFiles) {
            const templateFile = (_b = this.config.emails[templateName]) === null || _b === void 0 ? void 0 : _b.template;
            if (!templateFile) {
                return Promise.reject('No templates found for "' + templateName + '".');
            }
            templateFiles = [templateFile];
        }
        const readTemplates = templateFiles.map(t => (0, fs_1.readFileSync)(t, 'utf8'));
        for (let i = 0; i < templateFiles.length; i++) {
            if (!readTemplates[i]) {
                return Promise.reject('Failed to locate template file: ' + templateFiles[i]);
            }
        }
        const renderedTemplates = readTemplates.map(t => (0, ejs_1.render)(t, data));
        // form the email
        const subject = this.config.emails[templateName].subject;
        let formats = this.config.emails[templateName].formats;
        if (!formats) {
            const format = this.config.emails[templateName].format;
            if (!format) {
                return Promise.reject('No formats specified for: ' + templateName);
            }
            formats = [format];
        }
        if (formats.length !== renderedTemplates.length) {
            return Promise.reject('Different number of read templates and requested formats for template: ' +
                templateName);
        }
        let mailOptions = {
            from: this.config.mailer.fromEmail,
            to: recepient,
            subject: subject
        };
        if (this.config.mailer.messageConfig) {
            mailOptions = { ...this.config.mailer.messageConfig, ...mailOptions };
        }
        for (let i = 0; i < formats.length; i++) {
            mailOptions[formats[i]] = renderedTemplates[i];
        }
        if ((_c = this.config.testMode) === null || _c === void 0 ? void 0 : _c.debugEmail) {
            console.log(mailOptions);
        }
        // send the message
        return this.transporter.sendMail(mailOptions);
    }
}
exports.Mailer = Mailer;
